guides/hack/11-built-in-types/34-enum.md

```diff
@@ -0,0 +1,197 @@
+Use an enum (enumerated type) to create a set of named, constant, immutable values. 
+
+In Hack, enums are limited to `int` or `string` (as an [`Arraykey`](/hack/built-in-types/arraykey)), or other `enum` values.
+
+## Quickstart
+To access an enum's value, use its full name, as in `Colors::Blue` or `Permission::Read`.
+
+```Colors.hack no-auto-output
+enum Colors: int {
+  Red = 3;
+  Green = 5;
+  Blue = 10;
+  Default = 3; // duplicate value is okay
+}
+```
+
+```Permission.hack no-auto-output
+enum Permission: string {
+  Read = 'R';
+  Write = 'W';
+  Execute = 'E';
+  Delete = 'D';
+}
+```
+
+Additionally, by using the [`as`](/hack/expressions-and-operators/type-assertions#enforcing-types-with-as-and-as) operator to enforce type, you can initialize your enum with static expressions that reference other enum values.
+
+```BitFlags.hack no-auto-output
+enum BitFlags: int as int {
+  F1 = 1; // value 1
+  F2 = BitFlags::F1 << 1; // value 2
+  F3 = BitFlags::F2 << 1; // value 4
+  F4 = 4 + 3; // value 7
+}
+```
+
+## Full Example
+With an enum, we can create a placement-direction system with names like `Top`, `Bottom`, `Left`, `Right`, and `Center`, then direct output accordingly to write text to the top, bottom, left, right, or center of a window.
+
+```Positions.inc.hack no-auto-output
+enum Position: int {
+  Top = 0;
+  Bottom = 1;
+  Left = 2;
+  Right = 3;
+  Center = 4;
+}
+
+function writeText(string $text, Position $pos): void {
+  switch ($pos) {
+    case Position::Top:
+      // ...
+      break;
+    case Position::Center:
+      // ...
+      break;
+    case Position::Right:
+      // ...
+      break;
+    case Position::Left:
+      // ...
+      break;
+    case Position::Bottom:
+      // ...
+      break;
+  }
+}
+
+<<__EntryPoint>>
+function main(): void {
+  writeText("Hello", Position::Bottom);
+  writeText("Today", Position::Left);
+}
+```
+
+## Library Methods
+All enums implement these public static methods.
+
+### `getNames() / getValues()`
+Returns a map-like array of enum constant values and their names.
```

Comment:
It is true that all enums implement these two methods. An enum with duplicate values will throw an exception from getNames. This is done because the returned dict could not contain all members, since it is keyed by the values of the members. It might be handy to document this behavior here.

Reply:
I made this much clearer, thanks!

![image](https://user-images.githubusercontent.com/5179225/146616610-48d318a8-b9c9-4fd0-b3b3-22f01e93cbcf.png)

Reply:
I believe you may have flipped the examples around. `getNames()` returns `value` => `MemberName` and `getValues()` returns `MemberName` => `value`. 🙂

I have changed the `getNames()` description, as _Can not contain duplicate keys_ does not apply to `getValues()`. If you want, you may use my text verbatim, use elements, or not use it all.

## getValues() / getNames()

Returns a `dict` of enum constant values and their names.

 - `getValues()` returns a `dict` where the keys are the enum names and the values are the enum constant values.
   - In the example below, the keys/values would be: `"Top" => 0`, `"Bottom" => 1`, etc.
 - `getNames()` returns a `dict` which is the flipped result of `getValues()`\*; the keys are the enum constant values and the values are the enum's named constants.
   - Following the same example, the keys/values would be: `"0 => "Top"`, `1 => "Bottom"`

\* Dicts can not contain duplicate keys. `getNames()` can not return all members in enums with duplicate constant values. Calling `getNames()` on an enum with duplicate constant values will throw an `HH\InvariantException`, since the alternative would have been silently discarding members. If you want to silently discard members instead of throwing an exception, use `HH\Lib\Dict\flip(Position::getValues())`

```HACK
enum Position: int {
  Top = 0;
  Bottom = 1;
  Left = 2;
  Right = 3;
  Center = 4;
  Centre = 4;
}

<<__EntryPoint>>
function main(): void {
  \var_dump(Position::getNames());
}
```

```
'HH\InvariantException' with message
'Enum Position has overlapping values Center and Centre'
```

Reply:
oh, you're totally right! Thank you :-)

Reply:
Going to merge but I can always stealth edit. Fixed the mistake and wrote some more context on why `getNames()` presents an issue.

![image](https://user-images.githubusercontent.com/5179225/146627076-8e29cb44-ff5a-46dd-bc58-7d6c1ac231c1.png)